File: exam_frontend/core/js/create-exam.js
// DOM Elements
const uploadArea = document.getElementById('uploadArea');
const uploadSection = document.getElementById('uploadSection');
const fileInput = document.getElementById('fileInput');
const browseBtn = document.getElementById('browseBtn');
const fileInfo = document.getElementById('fileInfo');
const previewSection = document.getElementById('previewSection');
const previewTableBody = document.getElementById('previewTableBody');
const questionCount = document.getElementById('questionCount');
const submitBtn = document.getElementById('submitBtn');
const cancelBtn = document.getElementById('cancelBtn');
const progressContainer = document.getElementById('progressContainer');
const progressFill = document.getElementById('progressFill');
const progressText = document.getElementById('progressText');
const successMessage = document.getElementById('successMessage');
const errorMessage = document.getElementById('errorMessage');
const successText = document.getElementById('successText');
const errorText = document.getElementById('errorText');

// File info elements
const fileName = document.getElementById('fileName');
const fileSize = document.getElementById('fileSize');
const fileType = document.getElementById('fileType');
const fileModified = document.getElementById('fileModified');

// Global variables
let selectedFile = null;
let previewQuestionsList = [];

document.addEventListener('DOMContentLoaded', function() {
    if (typeof createParticles === 'function') {
        createParticles();
    }
    hidePreloader();
    
    // Check authentication
    checkAuth();
});

// Check if user is authenticated
function checkAuth() {
    const accessToken = CookieManager.get('access_token');
    if (!accessToken) {
        window.location.href = '/exam_frontend/core/auth.html';
    }
}

// Event Listeners
browseBtn.addEventListener('click', () => fileInput.click());
fileInput.addEventListener('change', handleFileSelect);
uploadArea.addEventListener('dragover', handleDragOver);
uploadArea.addEventListener('dragleave', handleDragLeave);
uploadArea.addEventListener('drop', handleDrop);
submitBtn.addEventListener('click', submitExam);
cancelBtn.addEventListener('click', resetForm);

// Drag and Drop Handlers
function handleDragOver(e) {
    e.preventDefault();
    uploadSection.classList.add('active');
}

function handleDragLeave(e) {
    e.preventDefault();
    uploadSection.classList.remove('active');
}

function handleDrop(e) {
    e.preventDefault();
    uploadSection.classList.remove('active');
    
    const files = e.dataTransfer.files;
    if (files.length > 0) {
        fileInput.files = files;
        handleFileSelect({ target: fileInput });
    }
}

// File Selection Handler
function handleFileSelect(e) {
    const file = e.target.files[0];
    if (!file) return;

    // Validate file type
    const validTypes = [
        'application/vnd.ms-excel',
        'application/vnd.openxmlformats-officedocument.spreadsheetml.sheet',
        'text/csv'
    ];
    
    const fileExtension = file.name.split('.').pop().toLowerCase();
    const isValidType = validTypes.includes(file.type) || 
                        ['xlsx', 'xls', 'csv'].includes(fileExtension);

    if (!isValidType) {
        showError('Please select a valid Excel or CSV file.');
        return;
    }

    // Store the selected file
    selectedFile = file;
    
    // Update file info
    updateFileInfo(file);
    
    // Show file preview prompt
    showInfo('File selected. Please fill in exam details below and preview the questions.');
    
    // Show exam details form
    showExamDetailsForm();
}

// Update File Information
function updateFileInfo(file) {
    fileName.textContent = file.name;
    fileSize.textContent = formatFileSize(file.size);
    fileType.textContent = file.type || `*.${file.name.split('.').pop()}`;
    fileModified.textContent = new Date(file.lastModified).toLocaleString();
    fileInfo.classList.add('show');
}

// Format File Size
function formatFileSize(bytes) {
    if (bytes === 0) return '0 Bytes';
    const k = 1024;
    const sizes = ['Bytes', 'KB', 'MB', 'GB'];
    const i = Math.floor(Math.log(bytes) / Math.log(k));
    return parseFloat((bytes / Math.pow(k, i)).toFixed(2)) + ' ' + sizes[i];
}

// Show Progress Bar
function showProgress() {
    progressContainer.classList.add('show');
    let progress = 0;
    const interval = setInterval(() => {
        progress += Math.random() * 10;
        if (progress >= 100) {
            progress = 100;
            clearInterval(interval);
        }
        progressFill.style.width = progress + '%';
        progressText.textContent = `Processing file... ${Math.round(progress)}%`;
    }, 200);
}

// Hide Progress Bar
function hideProgress() {
    progressContainer.classList.remove('show');
    progressFill.style.width = '0%';
}

// Show Exam Details Form
function showExamDetailsForm() {
    // Create form HTML if not exists
    let examDetailsForm = document.getElementById('examDetailsForm');
    if (!examDetailsForm) {
        examDetailsForm = document.createElement('div');
        examDetailsForm.id = 'examDetailsForm';
        examDetailsForm.className = 'exam-details-form';
        examDetailsForm.innerHTML = `
            <h3>Exam Details</h3>
            <form id="examForm">
                <div class="form-group">
                    <label for="examTitle">Exam Title *</label>
                    <input type="text" id="examTitle" required placeholder="e.g., Mathematics Grade 10">
                </div>
                
                <div class="form-group">
                    <label for="examCode">Exam Code *</label>
                    <input type="text" id="examCode" required placeholder="e.g., MATH101">
                </div>
                
                <div class="form-group">
                    <label for="examLevel">Academic Level *</label>
                    <select id="examLevel" required>
                        <option value="">Loading levels...</option>
                    </select>
                </div>
                
                <div class="form-group">
                    <label for="examDescription">Description</label>
                    <textarea id="examDescription" rows="3" placeholder="Optional exam description"></textarea>
                </div>
                
                <div class="form-row">
                    <div class="form-group">
                        <label for="examDuration">Duration (minutes) *</label>
                        <input type="number" id="examDuration" value="60" min="1" max="300" required>
                    </div>
                    
                    <div class="form-group">
                        <label for="passingScore">Passing Score (%) *</label>
                        <input type="number" id="passingScore" value="60" min="0" max="100" required>
                    </div>
                    
                    <div class="form-group">
                        <label for="maxAttempts">Max Attempts *</label>
                        <input type="number" id="maxAttempts" value="1" min="1" max="10" required>
                    </div>
                </div>
                
                <div class="form-checkboxes">
                    <label>
                        <input type="checkbox" id="randomizeQuestions" checked>
                        Randomize question order for each student
                    </label>
                    
                    <label>
                        <input type="checkbox" id="showCorrectAnswers" checked>
                        Show correct answers after exam completion
                    </label>
                    
                    <label>
                        <input type="checkbox" id="isActive">
                        Activate exam immediately (students can take it right away)
                    </label>
                </div>
                
                <button type="button" class="btn btn-primary" id="previewBtn">
                    <i class="fas fa-eye"></i> Preview Questions
                </button>
            </form>
        `;
        
        // Insert after file info
        fileInfo.parentNode.insertBefore(examDetailsForm, fileInfo.nextSibling);
    }
    
    examDetailsForm.classList.add('show');
    
    // Load levels
    loadLevels();
    
    // Add preview button listener
    document.getElementById('previewBtn').addEventListener('click', previewQuestions);
}

// Load Academic Levels
async function loadLevels() {
    try {
        const accessToken = CookieManager.get('access_token');
        const response = await fetch(`${CORE_URL}/levels/`, {
            method: 'GET',
            headers: {
                'Authorization': `Bearer ${accessToken}`,
                'Content-Type': 'application/json'
            }
        });
        
        const result = await response.json();
        
        if (response.ok && result.success) {
            const levelSelect = document.getElementById('examLevel');
            levelSelect.innerHTML = '<option value="">Select a level</option>';
            
            result.data.forEach(level => {
                const option = document.createElement('option');
                option.value = level.id;
                option.textContent = `${level.name} (${level.level_type})`;
                levelSelect.appendChild(option);
            });
        } else {
            showError('Failed to load academic levels');
        }
    } catch (error) {
        console.error('Error loading levels:', error);
        showError('Error loading academic levels');
    }
}

// Preview Questions (validate form and show preview)
async function previewQuestions() {
    // Validate form
    const examTitle = document.getElementById('examTitle').value.trim();
    const examCode = document.getElementById('examCode').value.trim();
    const examLevel = document.getElementById('examLevel').value;
    
    if (!examTitle || !examCode || !examLevel) {
        showError('Please fill in all required fields (Title, Code, and Level)');
        return;
    }
    
    if (!selectedFile) {
        showError('Please select a file first');
        return;
    }
    
    // Show progress
    showProgress();
    progressText.textContent = 'Reading and validating questions...';
    
    try {
        // Read and parse the file
        const questions = await parseExcelFile(selectedFile);
        
        hideProgress();
        
        if (questions && questions.length > 0) {
            previewQuestionsList = questions;
            displayPreviewQuestions(questions);
            previewSection.classList.add('show');
            showSuccess(`Successfully loaded ${questions.length} questions from file!`);
        } else {
            showError('No valid questions found in the file. Please check the file format.');
        }
    } catch (error) {
        hideProgress();
        console.error('Error parsing file:', error);
        showError('Error reading file: ' + error.message);
    }
}

// Parse Excel/CSV File
async function parseExcelFile(file) {
    return new Promise((resolve, reject) => {
        const reader = new FileReader();
        const fileExtension = file.name.split('.').pop().toLowerCase();
        
        reader.onload = (e) => {
            try {
                const data = e.target.result;
                let questions = [];
                
                if (fileExtension === 'csv') {
                    // Parse CSV
                    questions = parseCSV(data);
                } else if (fileExtension === 'xlsx' || fileExtension === 'xls') {
                    // Parse Excel using SheetJS
                    questions = parseExcel(data);
                }
                
                resolve(questions);
            } catch (error) {
                reject(error);
            }
        };
        
        reader.onerror = () => reject(new Error('Failed to read file'));
        
        if (fileExtension === 'csv') {
            reader.readAsText(file);
        } else {
            reader.readAsArrayBuffer(file);
        }
    });
}

// Parse Excel Data using SheetJS
function parseExcel(data) {
    try {
        // Check if XLSX library is loaded
        if (typeof XLSX === 'undefined') {
            throw new Error('Excel parser library not loaded. Please refresh the page.');
        }
        
        // Read the workbook
        const workbook = XLSX.read(data, { type: 'array' });
        
        // Get first sheet
        const firstSheetName = workbook.SheetNames[0];
        const worksheet = workbook.Sheets[firstSheetName];
        
        // Convert to JSON
        const jsonData = XLSX.utils.sheet_to_json(worksheet, { defval: '' });
        
        if (jsonData.length === 0) {
            throw new Error('No data found in Excel file');
        }
        
        // Normalize column names (lowercase and trim)
        const questions = jsonData.map(row => {
            const normalizedRow = {};
            Object.keys(row).forEach(key => {
                const normalizedKey = key.toLowerCase().trim().replace(/\s+/g, '_');
                normalizedRow[normalizedKey] = row[key];
            });
            return normalizedRow;
        });
        
        // Validate required columns
        const requiredCols = ['question_text', 'option_a', 'option_b', 'option_c', 'option_d', 'correct_answer'];
        const firstRow = questions[0] || {};
        const missingCols = requiredCols.filter(col => !(col in firstRow));
        
        if (missingCols.length > 0) {
            throw new Error(`Missing required columns: ${missingCols.join(', ')}`);
        }
        
        // Filter valid questions
        const validQuestions = questions.filter(q => 
            q.question_text && 
            q.option_a && 
            q.option_b && 
            q.option_c && 
            q.option_d && 
            q.correct_answer
        );
        
        if (validQuestions.length === 0) {
            throw new Error('No valid questions found. Please ensure all required fields are filled.');
        }
        
        return validQuestions;
    } catch (error) {
        throw new Error(`Excel parsing error: ${error.message}`);
    }
}

// Parse CSV Data
function parseCSV(csvText) {
    const lines = csvText.split('\n').filter(line => line.trim());
    if (lines.length < 2) return [];
    
    // Parse header
    const headers = lines[0].split(',').map(h => h.trim().toLowerCase().replace(/"/g, ''));
    
    // Required columns
    const requiredCols = ['question_text', 'option_a', 'option_b', 'option_c', 'option_d', 'correct_answer'];
    const missingCols = requiredCols.filter(col => !headers.includes(col));
    
    if (missingCols.length > 0) {
        throw new Error(`Missing required columns: ${missingCols.join(', ')}`);
    }
    
    // Parse rows
    const questions = [];
    for (let i = 1; i < lines.length; i++) {
        const values = parseCSVLine(lines[i]);
        if (values.length < headers.length) continue;
        
        const question = {};
        headers.forEach((header, index) => {
            question[header] = values[index] || '';
        });
        
        // Validate required fields
        if (question.question_text && question.option_a && question.option_b && 
            question.option_c && question.option_d && question.correct_answer) {
            questions.push(question);
        }
    }
    
    return questions;
}

// Parse CSV Line (handles quoted fields)
function parseCSVLine(line) {
    const result = [];
    let current = '';
    let inQuotes = false;
    
    for (let i = 0; i < line.length; i++) {
        const char = line[i];
        
        if (char === '"') {
            inQuotes = !inQuotes;
        } else if (char === ',' && !inQuotes) {
            result.push(current.trim().replace(/^"|"$/g, ''));
            current = '';
        } else {
            current += char;
        }
    }
    
    result.push(current.trim().replace(/^"|"$/g, ''));
    return result;
}

// Display Preview Questions
function displayPreviewQuestions(questions) {
    previewTableBody.innerHTML = '';
    
    if (!questions || questions.length === 0) {
        previewTableBody.innerHTML = `
            <tr>
                <td colspan="6" style="text-align: center; padding: 2rem;">
                    No questions to display
                </td>
            </tr>
        `;
        questionCount.textContent = '0';
        return;
    }
    
    questions.forEach((q, index) => {
        const row = document.createElement('tr');
        const options = [q.option_a, q.option_b, q.option_c, q.option_d].filter(opt => opt).join(', ');
        const correctAnswer = q.correct_answer ? q.correct_answer.toUpperCase() : '-';
        const difficulty = q.difficulty || 'medium';
        const category = q.category || 'General';
        
        row.innerHTML = `
            <td>${index + 1}</td>
            <td>${q.question_text.substring(0, 100)}${q.question_text.length > 100 ? '...' : ''}</td>
            <td style="font-size: 0.85rem;">${options.substring(0, 50)}${options.length > 50 ? '...' : ''}</td>
            <td><strong style="color: var(--primary-color);">${correctAnswer}</strong></td>
            <td><span class="file-type-badge">${difficulty}</span></td>
            <td>${category}</td>
        `;
        previewTableBody.appendChild(row);
    });
    
    questionCount.textContent = questions.length;
}

// Submit Exam to Backend
async function submitExam() {
    if (!selectedFile) {
        showError('Please select a file first');
        return;
    }
    
    // Get form values
    const examTitle = document.getElementById('examTitle')?.value.trim();
    const examCode = document.getElementById('examCode')?.value.trim();
    const examLevel = document.getElementById('examLevel')?.value;
    const examDescription = document.getElementById('examDescription')?.value.trim() || '';
    const examDuration = document.getElementById('examDuration')?.value || 60;
    const passingScore = document.getElementById('passingScore')?.value || 60;
    const maxAttempts = document.getElementById('maxAttempts')?.value || 1;
    const randomizeQuestions = document.getElementById('randomizeQuestions')?.checked || false;
    const showCorrectAnswers = document.getElementById('showCorrectAnswers')?.checked || false;
    const isActive = document.getElementById('isActive')?.checked || false;
    
    // Validate required fields
    if (!examTitle || !examCode || !examLevel) {
        showError('Please fill in all required fields (Title, Code, and Level)');
        return;
    }
    
    showProgress();
    progressText.textContent = 'Uploading and creating exam...';
    
    try {
        // Create FormData
        const formData = new FormData();
        formData.append('file', selectedFile);
        formData.append('title', examTitle);
        formData.append('code', examCode);
        formData.append('level_id', examLevel);
        formData.append('description', examDescription);
        formData.append('duration_minutes', examDuration);
        formData.append('passing_score', passingScore);
        formData.append('max_attempts', maxAttempts);
        formData.append('randomize_questions', randomizeQuestions);
        formData.append('show_correct_answers', showCorrectAnswers);
        formData.append('is_active', isActive);
        
        // Get access token
        const accessToken = CookieManager.get('access_token');
        
        // Send request to backend
        const response = await fetch(`${CORE_URL}/exams/create/`, {
            method: 'POST',
            headers: {
                'Authorization': `Bearer ${accessToken}`
            },
            body: formData
        });
        
        const result = await response.json();
        
        hideProgress();
        
        if (response.ok && result.success) {
            const examData = result.data.exam_module;
            const questionsData = result.data.questions;
            const uploadData = result.data.bulk_upload;
            
            // Show success message with details
            let successMsg = `Exam "${examData.title}" created successfully with ${examData.total_questions} questions!`;
            
            if (uploadData.failed_records > 0) {
                successMsg += ` Note: ${uploadData.failed_records} questions failed validation.`;
            }
            
            showSuccess(successMsg);
            
            // Display actual questions in preview
            displayCreatedQuestions(questionsData);
            
            // Reset form after 3 seconds
            setTimeout(() => {
                resetForm();
                // Redirect to dashboard or exam list
                // window.location.href = '/exam_frontend/core/admin-dashboard.html';
            }, 3000);
        } else {
            showError(result.message || 'Failed to create exam. Please check your file format and try again.');
        }
    } catch (error) {
        hideProgress();
        console.error('Error creating exam:', error);
        showError('Network error. Please check your connection and try again.');
    }
}

// Display Created Questions
function displayCreatedQuestions(questions) {
    previewTableBody.innerHTML = '';
    
    if (!questions || questions.length === 0) {
        previewTableBody.innerHTML = `
            <tr>
                <td colspan="6" style="text-align: center; padding: 2rem;">
                    No questions to display
                </td>
            </tr>
        `;
        return;
    }
    
    questions.forEach((q, index) => {
        const row = document.createElement('tr');
        row.innerHTML = `
            <td>${index + 1}</td>
            <td>${q.question_text}</td>
            <td>-</td>
            <td>-</td>
            <td><span class="file-type-badge">${q.difficulty}</span></td>
            <td>${q.category || 'General'}</td>
        `;
        previewTableBody.appendChild(row);
    });
    
    questionCount.textContent = questions.length;
}

// Show Success Message
function showSuccess(message) {
    successText.textContent = message;
    successMessage.classList.add('show');
    errorMessage.classList.remove('show');
    
    setTimeout(() => {
        successMessage.classList.remove('show');
    }, 5000);
}

// Show Error Message
function showError(message) {
    errorText.textContent = message;
    errorMessage.classList.add('show');
    successMessage.classList.remove('show');
    
    setTimeout(() => {
        errorMessage.classList.remove('show');
    }, 5000);
}

// Show Info Message
function showInfo(message) {
    // You can create a new info message element or reuse success
    successText.textContent = message;
    successMessage.classList.add('show');
    errorMessage.classList.remove('show');
    
    setTimeout(() => {
        successMessage.classList.remove('show');
    }, 5000);
}

// Reset Form
function resetForm() {
    fileInput.value = '';
    selectedFile = null;
    fileInfo.classList.remove('show');
    previewSection.classList.remove('show');
    successMessage.classList.remove('show');
    errorMessage.classList.remove('show');
    uploadSection.classList.remove('active');
    
    // Reset exam details form
    const examDetailsForm = document.getElementById('examDetailsForm');
    if (examDetailsForm) {
        examDetailsForm.remove();
    }
}

File: exam_backend/apps/core/BLL/Commands/create_exam.py
"""
Create Exam Command - SmartExam System
========================================

This module handles exam creation with bulk question upload from Excel/CSV files.

Features:
- Parse Excel (.xlsx, .xls) and CSV files
- Validate question data
- Create ExamModule and Questions in bulk
- Track upload history
- Provide detailed error reporting
"""

import pandas as pd
from io import BytesIO
from django.db import transaction
from utils.base_result import BaseResult
from apps.core.models import ExamModule, Question, Level, BulkUpload
import traceback


class CreateExamCommand:
    """
    Command to create exam with questions from uploaded file
    """
    
    def __init__(self, data, file, user):
        """
        Initialize command with exam data and uploaded file
        
        Args:
            data (dict): Exam details (title, code, level_id, etc.)
            file (File): Uploaded Excel/CSV file
            user (User): Admin user creating the exam
        """
        self.data = data
        self.file = file
        self.user = user
        self.result = BaseResult()
        self.questions_data = []
        self.errors = []
    
    def execute(self):
        """
        Main execution method
        
        Returns:
            BaseResult: Contains exam_module, questions, and upload tracking info
        """
        try:
            print(f"[DEBUG] Starting exam creation: {self.data.get('title')}")
            
            # Step 1: Parse the file
            if not self._parse_file():
                print(f"[ERROR] File parsing failed: {self.result.message}")
                return self.result
            
            print(f"[DEBUG] Successfully parsed {len(self.questions_data)} questions")
            
            # Step 2: Validate questions data
            if not self._validate_questions():
                print(f"[ERROR] Question validation failed: {self.result.message}")
                return self.result
            
            print(f"[DEBUG] Validated {len(self.questions_data)} valid questions")
            
            # Step 3: Create exam and questions in transaction
            with transaction.atomic():
                # Create exam module
                exam_module = self._create_exam_module()
                if not exam_module:
                    print(f"[ERROR] Exam module creation failed")
                    return self.result
                
                print(f"[DEBUG] Created exam module: {exam_module.code}")
                
                # Create questions
                questions = self._create_questions(exam_module)
                print(f"[DEBUG] Created {len(questions)} questions")
                
                # Create bulk upload record
                bulk_upload = self._create_bulk_upload_record(
                    exam_module, 
                    len(questions)
                )
                
                self.result.success = True
                self.result.message = f"Exam '{exam_module.title}' created successfully with {len(questions)} questions"
                self.result.data = {
                    'exam_module': {
                        'id': exam_module.id,
                        'title': exam_module.title,
                        'code': exam_module.code,
                        'total_questions': len(questions),
                        'total_marks': exam_module.total_marks,
                    },
                    'questions': [
                        {
                            'id': q.id,
                            'question_text': q.question_text,
                            'difficulty': q.difficulty,
                            'category': q.category,
                            'marks': q.marks,
                        }
                        for q in questions
                    ],
                    'bulk_upload': {
                        'id': bulk_upload.id,
                        'total_records': bulk_upload.total_records,
                        'successful_records': bulk_upload.successful_records,
                        'failed_records': bulk_upload.failed_records,
                        'status': bulk_upload.status,
                    }
                }
                
        except Exception as e:
            print(f"[ERROR] Exception in execute: {str(e)}")
            traceback.print_exc()
            self.result.success = False
            self.result.message = f"Error creating exam: {str(e)}"
            self.result.status_code = 500
        
        return self.result
    
    
    def _parse_file(self):
        """
        Parse Excel or CSV file and extract questions data
        
        Returns:
            bool: True if parsing successful, False otherwise
        """
        try:
            file_name = self.file.name.lower()
            print(f"[DEBUG] Parsing file: {file_name}")
            
            # Read file content into BytesIO
            file_content = self.file.read()
            file_io = BytesIO(file_content)
            
            # Read file based on extension
            if file_name.endswith('.csv'):
                df = pd.read_csv(file_io)
                print(f"[DEBUG] CSV file loaded, shape: {df.shape}")
            elif file_name.endswith(('.xlsx', '.xls')):
                df = pd.read_excel(file_io, engine='openpyxl')
                print(f"[DEBUG] Excel file loaded, shape: {df.shape}")
            else:
                self.result.success = False
                self.result.message = "Unsupported file format. Please upload Excel (.xlsx, .xls) or CSV (.csv) file"
                self.result.status_code = 400
                return False
            
            # Check if dataframe is empty
            if df.empty:
                self.result.success = False
                self.result.message = "The uploaded file is empty"
                self.result.status_code = 400
                return False
            
            print(f"[DEBUG] Original columns: {df.columns.tolist()}")
            
            # Standardize column names (lowercase and strip spaces, replace spaces with underscores)
            df.columns = df.columns.str.lower().str.strip().str.replace(' ', '_')
            print(f"[DEBUG] Normalized columns: {df.columns.tolist()}")
            
            # Check required columns
            required_columns = [
                'question_text', 'option_a', 'option_b', 'option_c', 
                'option_d', 'correct_answer'
            ]
            missing_columns = [col for col in required_columns if col not in df.columns]
            
            if missing_columns:
                self.result.success = False
                self.result.message = f"Missing required columns: {', '.join(missing_columns)}. Found columns: {', '.join(df.columns.tolist())}"
                self.result.status_code = 400
                return False
            
            # Convert dataframe to list of dictionaries, replacing NaN with empty strings
            self.questions_data = df.fillna('').to_dict('records')
            print(f"[DEBUG] Extracted {len(self.questions_data)} rows from file")
            
            return True
            
        except Exception as e:
            print(f"[ERROR] File parsing exception: {str(e)}")
            traceback.print_exc()
            self.result.success = False
            self.result.message = f"Error parsing file: {str(e)}"
            self.result.status_code = 400
            return False
    
    
    def _validate_questions(self):
        """
        Validate all questions data
        
        Returns:
            bool: True if all questions are valid, False otherwise
        """
        if not self.questions_data:
            self.result.success = False
            self.result.message = "No questions found in the uploaded file"
            self.result.status_code = 400
            return False
        
        valid_questions = []
        
        for idx, question in enumerate(self.questions_data, start=1):
            errors_for_row = []
            
            # Validate required fields
            if not str(question.get('question_text', '')).strip():
                errors_for_row.append(f"Row {idx}: Question text is required")
            
            # Validate options
            for opt in ['option_a', 'option_b', 'option_c', 'option_d']:
                if not str(question.get(opt, '')).strip():
                    errors_for_row.append(f"Row {idx}: {opt.replace('_', ' ').title()} is required")
            
            # Validate correct answer
            correct_answer = str(question.get('correct_answer', '')).upper().strip()
            if correct_answer not in ['A', 'B', 'C', 'D']:
                errors_for_row.append(f"Row {idx}: Correct answer must be A, B, C, or D (found: '{correct_answer}')")
            
            # If there are errors for this row, add them and skip
            if errors_for_row:
                self.errors.extend(errors_for_row)
                continue
            
            # Set defaults for optional fields
            question['difficulty'] = str(question.get('difficulty', 'medium')).lower().strip()
            if question['difficulty'] not in ['easy', 'medium', 'hard']:
                question['difficulty'] = 'medium'
            
            question['category'] = str(question.get('category', 'General')).strip() or 'General'
            
            # Handle marks
            try:
                marks_value = question.get('marks', 1)
                question['marks'] = int(float(marks_value)) if marks_value else 1
            except:
                question['marks'] = 1
            
            question['explanation'] = str(question.get('explanation', '')).strip() or None
            
            # Handle order
            try:
                order_value = question.get('order', 0)
                question['order'] = int(float(order_value)) if order_value else 0
            except:
                question['order'] = 0
            
            # Ensure correct_answer is uppercase
            question['correct_answer'] = correct_answer
            
            valid_questions.append(question)
        
        # Check if we have any valid questions
        if not valid_questions:
            self.result.success = False
            self.result.message = "No valid questions found"
            self.result.data = {'errors': self.errors}
            self.result.status_code = 400
            return False
        
        # Update questions data with only valid questions
        self.questions_data = valid_questions
        
        # Include warnings if some questions were skipped
        if self.errors:
            print(f"[WARNING] {len(self.errors)} validation errors found")
            for error in self.errors:
                print(f"  - {error}")
            self.result.data = {
                'warnings': self.errors,
                'valid_questions': len(valid_questions),
                'invalid_questions': len(self.errors)
            }
        
        return True
    
    def _create_exam_module(self):
        """
        Create ExamModule instance
        
        Returns:
            ExamModule: Created exam module instance
        """
        try:
            level = Level.objects.get(id=self.data['level_id'], is_active=True)
            
            exam_module = ExamModule.objects.create(
                title=self.data['title'],
                description=self.data.get('description', ''),
                code=self.data['code'],
                level=level,
                duration_minutes=self.data.get('duration_minutes', 60),
                passing_score=self.data.get('passing_score', 60),
                max_attempts=self.data.get('max_attempts', 1),
                randomize_questions=self.data.get('randomize_questions', True),
                show_correct_answers=self.data.get('show_correct_answers', True),
                is_active=self.data.get('is_active', False),
            )
            
            return exam_module
            
        except Level.DoesNotExist:
            self.result.success = False
            self.result.message = "Invalid level ID"
            self.result.status_code = 400
            return None
        except Exception as e:
            self.result.success = False
            self.result.message = f"Error creating exam module: {str(e)}"
            self.result.status_code = 500
            return None
    
    def _create_questions(self, exam_module):
        """
        Create Question instances in bulk
        
        Args:
            exam_module (ExamModule): Exam module to associate questions with
        
        Returns:
            list: List of created Question instances
        """
        questions = []
        
        for question_data in self.questions_data:
            question = Question(
                exam_module=exam_module,
                question_text=question_data['question_text'],
                option_a=question_data['option_a'],
                option_b=question_data['option_b'],
                option_c=question_data['option_c'],
                option_d=question_data['option_d'],
                correct_answer=str(question_data['correct_answer']).upper().strip(),
                difficulty=question_data['difficulty'],
                category=question_data['category'],
                marks=question_data['marks'],
                explanation=question_data.get('explanation'),
                order=question_data.get('order', 0),
                is_active=True
            )
            questions.append(question)
        
        # Bulk create questions
        created_questions = Question.objects.bulk_create(questions)
        
        return created_questions
    
    def _create_bulk_upload_record(self, exam_module, questions_count):
        """
        Create BulkUpload tracking record
        
        Args:
            exam_module (ExamModule): Created exam module
            questions_count (int): Number of questions created
        
        Returns:
            BulkUpload: Created bulk upload record
        """
        bulk_upload = BulkUpload.objects.create(
            upload_type='questions',
            file=self.file,
            uploaded_by=self.user,
            total_records=len(self.questions_data) + len(self.errors),
            successful_records=questions_count,
            failed_records=len(self.errors),
            status='completed' if not self.errors else 'completed',
            error_log='\n'.join(self.errors) if self.errors else None
        )
        
        return bulk_upload
